ngsPETSc/utils/firedrake/adaptive_hierarchy.py

```diff
@@ -0,0 +1,331 @@
+        return refined_mesh
+
+    def split_function(self, u, child=True):
+        """
+        Split input function across submeshes
+        """
+        V = u.function_space()
+        full_mesh = V.mesh()
+        _, level = get_level(full_mesh)
+
+        ind = 1 if child else 0
+        hierarchy_dict = self.submesh_hierarchies[int(level) - ind]
+        parent_mesh = hierarchy_dict[[*hierarchy_dict][0]].meshes[ind].submesh_parent
+        parent_space = V.reconstruct(parent_mesh)
+        u_corr_space = Function(parent_space, val=u.dat)
+        key = (u, child)
+        try:
+            split_functions = self.split_cache[key]
+        except KeyError:
+            split_functions = self.split_cache.setdefault(key, {})
+
+        for i in hierarchy_dict:
+            try:
+                f = split_functions[i].zero()
+            except KeyError:
+                V_split = V.reconstruct(mesh=hierarchy_dict[i].meshes[ind])
+                assert (
+                    V_split.mesh().submesh_parent
+                    == u_corr_space.function_space().mesh()
+                )
+                f = split_functions.setdefault(i, Function(V_split, name=str(i)))
+
+            f.assign(u_corr_space)
+        return split_functions
+
+    def use_weight(self, V, child):
+        """
+        Counts DoFs across submeshes, computes partition of unity
+        """
+        w = Function(V).assign(1)
+        splits = self.split_function(w, child)
+
+        self.recombine(splits, w, child)
+        with w.dat.vec as wvec:
+            wvec.reciprocal()
+        return w
+
+    def recombine(self, split_funcs, f, child=True):
+        """
+        Recombines functions on submeshes back full mesh
+        """
+        V = f.function_space()
+        f.zero()
+        parent_mesh = (
+            split_funcs[[*split_funcs][0]].function_space().mesh().submesh_parent
+        )
+        V_label = V.reconstruct(mesh=parent_mesh)
+        if isinstance(f, Function):
+            f_label = Function(V_label, val=f.dat)
+        elif isinstance(f, Cofunction):
+            f_label = Cofunction(V_label, val=f.dat)
+
+        for split_label, val in split_funcs.items():
+            assert val.function_space().mesh().submesh_parent == parent_mesh
+            if child:
+                split_label = int("10" + str(split_label))
+            if isinstance(f_label, Function):
+                f_label.assign(val, allow_missing_dofs=True)
+            else:
+                curr = Function(f_label.function_space()).assign(
+                    val, allow_missing_dofs=True
+                )
+                f_label.assign(f_label + curr)  # partition of unity for restriction
+        return f
+
+
+def get_c2f_f2c_fd(mesh, coarse_mesh):
+    """
+    Construct coarse->fine and fine->coarse relations by mapping netgen elements to firedrake ones
+    """
+    _ = FunctionSpace(mesh, "DG", 0)
+    _ = FunctionSpace(coarse_mesh, "DG", 0)
```

Comment:
Why is this needed? Do we need to call `mesh.init()`? That behaivour has been recently changed, so it might be able to work fine without these calls.
```suggestion
```